src/main/java/org/springframework/shell/table/Table.java

```diff
@@ -0,0 +1,389 @@
+
+			}
+		}
+
+
+		cellWidths = computeActualColumnWidths(widthAvailableForContents, minCellWidths, maxCellWidths);
+		// Now that widths are known, apply wrapping & render
+		for (int row = 0; row < rows; row++) {
+			for (int column = 0; column < columns; column++) {
+				subLines[row][column] = getWrapper(row, column).wrap(subLines[row][column], cellWidths[column]);
+				cellHeights[row] = Math.max(cellHeights[row], subLines[row][column].length);
+			}
+			for (int column = 0; column < columns; column++) {
+				for (Map.Entry<CellMatcher, AlignmentStrategy> kv : aligners.entrySet()) {
+					if (kv.getKey().matches(row, column, model)) {
+						subLines[row][column] = kv.getValue().align(subLines[row][column], cellWidths[column], cellHeights[row]);
+					}
+				}
+			}
+		}
+
+
+		for (int row = 0; row < rows; row++) {
+
+			// TOP CELL BORDER
+			int before = result.length();
+			for (int column = 0; column < columns; column++) {
+				borders.paintCorner(row, column, result);
+				borders.paintHorizontal(row, column, cellWidths[column], result);
+			}
+			borders.paintCorner(row, columns, result);
+			if (result.length() > before) {
+				result.append('\n');
+			}
+
+			for (int subRow = 0; subRow < cellHeights[row]; subRow++) {
+				for (int column = 0; column < columns; column++) {
+					// LEFT CELL BORDER
+					borders.paintVertical(row, column, result);
+					String[] lines = subLines[row][column];
+					result.append(lines[subRow]);
+				}
+				// TABLE RIGHT BORDER
+				borders.paintVertical(row, columns, result);
+				result.append("\n");
+			}
+		}
+
+		// TABLE BOTTOM BORDER
+		int before = result.length();
+		for (int column = 0; column < columns; column++) {
+			borders.paintCorner(rows, column, result);
+			borders.paintHorizontal(rows, column, cellWidths[column], result);
+		}
+
+		// TABLE BOTTOM RIGHT CORNER
+		borders.paintCorner(rows, columns, result);
+		if (result.length() > before) {
+			result.append('\n');
+		}
+		return result.toString();
+	}
+
+	private int[] computeActualColumnWidths(int availableWidth, int[] minCellWidths, int[] maxCellWidths) {
+
+		int[] cellWidths;
+		int minTableWidth = 0, maxTableWidth = 0;
+		for (int column = 0; column < columns; column++) {
+			minTableWidth += minCellWidths[column];
+			maxTableWidth += maxCellWidths[column];
+		}
+
+		// Can use max desired width
+		if (maxTableWidth <= availableWidth) {
+			cellWidths = maxCellWidths;
+		} // will overflow
+		else if (minTableWidth >= availableWidth) {
+			cellWidths = minCellWidths;
+		} // Redistribute nicely
+		else {
+			int W = availableWidth - minTableWidth;
+			int D = maxTableWidth - minTableWidth;
+			cellWidths = new int[columns];
+			for (int column = 0; column < columns; column++) {
+				cellWidths[column] = minCellWidths[column] + W * (maxCellWidths[column] - minCellWidths[column]) / D;
+			}
+		}
+		// TODO: handle rounding error?
+		return cellWidths;
+	}
+
+	public Table format(CellMatcher cells, Formatter formatter) {
```

Comment:
Also a builder class looks like it could hide the repetition of 

```
		table.format(CellMatchers.ofType(Map.class), new MapFormatter(" = "));
		table.align(CellMatchers.ofType(Map.class), new KeyValueHorizontalAligner("="));
		table.size(CellMatchers.ofType(Map.class), new KeyValueSizeConstraints(" = "));
		table.wrap(CellMatchers.ofType(Map.class), new KeyValueTextWrapper());```

passing in the same `CellMatchers` instance over and over again.